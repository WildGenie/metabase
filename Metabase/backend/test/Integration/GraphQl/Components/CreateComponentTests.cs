using System.IO;
using System.Threading.Tasks;
using Snapshooter.Xunit;
using Xunit;
using Metabase.GraphQl.Components;
using FluentAssertions;
using Snapshooter;

namespace Metabase.Tests.Integration.GraphQl.Components
{
    [Collection(nameof(Data.Component))]
    public sealed class CreateComponentTests
      : ComponentIntegrationTests
    {
        [Fact]
        public async Task AnonymousUser_IsAuthenticationError()
        {
            // Act
            var response =
             await UnsuccessfullyQueryGraphQlContentAsString(
                File.ReadAllText("Integration/GraphQl/Components/CreateComponent.graphql"),
                variables: MinimalComponentInput
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [Fact]
        public async Task AnonymousUser_CannotCreateComponent()
        {
            // Act
            await UnsuccessfullyQueryGraphQlContentAsString(
                File.ReadAllText("Integration/GraphQl/Components/CreateComponent.graphql"),
                variables: MinimalComponentInput
                ).ConfigureAwait(false);
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [Theory]
        [MemberData(nameof(EnumerateComponentInputs))]
        public async Task LoggedInUser_IsSuccess(
            string key,
            CreateComponentInput input
        )
        {
            // Arrange
            await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            // Act
            var response = await CreateComponent(input).ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                SnapshotNameExtension.Create(key),
                matchOptions => matchOptions.Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.createComponent.component.id").Should().NotBeNullOrWhiteSpace()
                 )
                );
        }

        [Theory]
        [MemberData(nameof(EnumerateComponentInputs))]
        public async Task LoggedInUser_CreatesComponent(
            string key,
            CreateComponentInput input
        )
        {
            // Arrange
            await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            // Act
            var componentId =
                ExtractString(
                    "$.data.createComponent.component.id",
                    await CreateComponentAsJson(input).ConfigureAwait(false)
                );
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                SnapshotNameExtension.Create(key),
                matchOptions => matchOptions.Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.components.edges[*].node.id").Should().Be(componentId)
                 )
                );
        }
    }
}