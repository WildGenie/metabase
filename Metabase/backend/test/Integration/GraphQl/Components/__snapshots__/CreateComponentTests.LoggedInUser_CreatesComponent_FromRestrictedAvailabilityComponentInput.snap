{"data":{"components":{"edges":[{"node":{"id":"Q29tcG9uZW50CmdiYzZlNGM5ZDM4Y2M0MWJjODllM2Y2MjkxNmIyYmI1Yg==","name":"Component D","abbreviation":"C!D","description":"Whatever component!","availability":{"from":"2019-04-03T00:00:00.000Z","to":null},"categories":["LAYER","MATERIAL","UNIT"]}}]}}}
